### ISSUE TITLE
add android and iOS to CI

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/RFCs#363]
## proposal
improve nim support for android and ios by testing for those platforms in CI.

These are important platforms, at least compared to netbsd, freebsd, openbsd.

### nim docs
* https://nim-lang.github.io/Nim/nimc.html#crossminuscompilation-for-android

### approach 1: docker
one approach could be docker, eg https://andresand.medium.com/android-emulator-on-docker-container-f20c49b129ef (see also https://github.com/nim-lang/Nim/pull/17424 which aims to run some pipelines via docker in our CI). This might be easier to test locally.

https://hub.docker.com/r/thyrlian/android-sdk/
See also https://github.com/budtmo/docker-android

### approach 2: azure pipelines
see https://docs.microsoft.com/en-us/azure/devops/pipelines/ecosystems/android?view=azure-devops

### option 3: github actions
see https://medium.com/google-developer-experts/github-actions-for-android-developers-6b54c8a32f55

### nim docs
* https://nim-lang.github.io/Nim/nimc.html#crossminuscompilation-for-ios

### option 1: docker
apparently, it's still possible:
* https://hub.docker.com/r/quamotion/appium-docker-ios
* https://github.com/quamotion/docker
* https://github.com/jkingyens/docker4xcode (or fork https://github.com/matsawyer/ios-emulator)

### option 2: azure pipelines
* https://docs.microsoft.com/en-us/azure/devops/pipelines/ecosystems/xcode?view=azure-devops
* https://medium.com/objectsharp/building-ios-application-using-azure-pipelines-92918080af9
* https://levelup.gitconnected.com/a-comprehensive-guide-to-use-azure-ci-cd-for-ios-apps-61d50ab9d077

### option 3: github actions
see https://engineering.talkdesk.com/test-and-deploy-an-ios-app-with-github-actions-44de9a7dcef6
example: https://github.com/Apple-Actions/Example-iOS

## somewhat related links
* [nim android tutorial - Nim forum](https://forum.nim-lang.org/t/5152#32331)
* [Nim for mobile - Nim forum](https://forum.nim-lang.org/t/7251)
* [iffy/wiish: Why Is It So Hard (to make a cross-platform app)?](https://github.com/iffy/wiish)
* [treeform/glfm: Wrapper of GLFM (OpenGL ES and input for iOS and Android) library for Nim.](https://github.com/treeform/glfm)
* [Native GUI development for MacOS - Nim forum](https://forum.nim-lang.org/t/7321#46530)
* [yglukhov/darwin: Bindings for MacOS/iOS frameworks](https://github.com/yglukhov/darwin)
* [SIGSEGV on Android - Nim forum](https://forum.nim-lang.org/t/7026#44179)
* [iOS compilation error - Nim forum](https://forum.nim-lang.org/t/6525#42126)
* [How to get Nim running on iOS and Android using GLFM. - Nim forum](https://forum.nim-lang.org/t/5197#37068)
* [Add -d:androidNDK to fix echo on Android NDK builds. by treeform · Pull Request #12203 · nim-lang/Nim](https://github.com/nim-lang/Nim/pull/12203) (relevant info here: https://github.com/nim-lang/Nim/pull/12203#issuecomment-532010158)
* [Using Nim to create an Xcode framework - Nim forum](https://forum.nim-lang.org/t/5268#33157)

help welcome @treeform, @alaviss etc

**Comment by ArMour85** [comment]
I agree that the CI should test more CPU architectures and we should already start with ARM32 and ARM64. Currently only AMD64 architecture is tested in the CI and now we can see that ARM processors become certainly the new standard for laptops (hello Apple M1 and Chromebook) but also lot of micro PC like Raspberry Pi.

If it is technically and financially possible (I don't know if it's expensive to use VM on ARM with Azure or another else) we should add the following platforms to the CI:
- ARM32/64 Linux
- MacOS M1

Furthermore, the next releases could be offered not only for AMD64 but for all these platforms.

**Comment by timotheecour** [comment]
CI for a given platform doesn't need to run on each PR push, nor even on each merge, that can be easily customized on a per-platform basis; the tricky part is to supporting a given platform in CI in the 1st place, the scaling aspect can be dealt with easily.

**Comment by ArMour85** [comment]
Yes, others platforms can be tested in CI in the pre-release stage for example. Run CI for all will be very long without any real benefit.

**Comment by alaviss** [comment]
> ARM32/64 Linux

I already have that setup actually: https://github.com/nim-lang/Nim/pull/16396

Let me know if there is interest in getting this updated. It's just sitting there right now because I don't see any interest for it in the community.

**Comment by Araq** [comment]
I'm interested but I don't review "Drafts", there are too many of them. Please un-draft it.

**Comment by al6x** [comment]
I think most people meant different thing by saying "support for Androind".

It could be a Nim repo, say `nim-androind-todo-example`, you clone it to your machine, run `build.sh` and get `todo.pkg` then you copy it to your Android phone, and see [something like this](http://svelte-todomvc.surge.sh/).
